### PR TITLE
Fix #388: off by one bug where wrong command is returned

### LINKED ISSUE #388: CommandHistory: First up arrow key press will go to second most recent command

### PR DESCRIPTION
**Pull request by AllardQuek**
- Keep track of whether the last command was valid or not

**Nuances:** 
- if valid, we should return the *current command* (i.e. the latest command entered), since on the UI the text field is reset to be empty after entering the command
- if invalid, we should return the previous command since the UI will display the invalid command in red to show that it is invalid

### CODE REVIEW

**Comment by shermannws** [comment]
Everytime the navigation goes to the oldest or most recent command, a double key press is required in order to retrieve the next comand.


e.g.
When counter is at 0 (oldest command), when `getNextCommand()` is executed, it will return the command again at index 0 before incrementing the counter.

File: src/main/java/seedu/programmer/ui/CommandBox.java [hunk 1]
```diff
@@ -25,6 +25,8 @@
     private final CommandExecutor commandExecutor;
     private final CommandHistory commandHistory = new CommandHistory();
 
+    private boolean lastCommandIsValid = false;
```
Comment (shermannws) [inline src/main/java/seedu/programmer/ui/CommandBox.java:28]: i think can rename to `isLastCommandValid` ?

Comment (AllardQuek) [reply src/main/java/seedu/programmer/ui/CommandBox.java:28]: Would personally prefer writing `if lastCommandIsValid` over `if isLastCommandValid`, but I can rename to be more in line with the module expectations. Good catch 😃

File: src/main/java/seedu/programmer/ui/CommandBox.java [hunk 2]
```diff
@@ -67,16 +71,25 @@ private void handleKeyPressed(KeyEvent event) {
...
 
-        if (downAndAtLastCommand || upAndAtFirstCommand) {
-            logger.info("We are already at the oldest or newest command -> show current command");
+        if (upAndAtLastCommand && lastCommandIsValid) {
+            // Remember the text field resets to "" after each valid command
+            commandTextField.setText(commandHistory.getCurrentCommand());
+            lastCommandIsValid = false;
+        } else if (downAndAtLastCommand && lastCommandIsValid) {
+            logger.info("We are already at the newest command -> show current command");
+            commandTextField.setText("");
```
Comment (shermannws) [inline src/main/java/seedu/programmer/ui/CommandBox.java:89]: ^there is still a bug now, to replicate error:
1. `show 1`
2. `show 2`
3. `show 3`
4. press down once --> nothing happens
5. press down again --> current command appears
6. then works as per normal

Suggestion:
1. Change line 89 to this line `commandTextField.setText(commandHistory.getCurrentCommand());`

File: src/main/java/seedu/programmer/ui/CommandBox.java (hunk 1 above)
Comment (shermannws) [reply src/main/java/seedu/programmer/ui/CommandBox.java:28]: I think would be good to follow closely with the module expectations:
How about another suggestion: `wasLastCommandValid`?

I think can also add a comment here to define this boolean flag:
"// boolean flag to keep track if valid command has just been executed and no keyboard press has been executed yet

File: src/main/java/seedu/programmer/ui/CommandBox.java (hunk 2 above)
Comment (AllardQuek) [reply src/main/java/seedu/programmer/ui/CommandBox.java:89]: Good catch! How about removing line 90? Seems to work for me

File: src/main/java/seedu/programmer/ui/CommandBox.java (hunk 1 above)
Comment (AllardQuek) [reply src/main/java/seedu/programmer/ui/CommandBox.java:28]: Good idea, I'll add a comment and let's use `isLastCommandValid` for the variable name

**Review by shermannws — APPROVED** [review]
LGTM! Thanks a lot!

**Pull request merged**